client/base_client.go

```diff
@@ -32,32 +32,110 @@ import (
 	"github.com/tikv/pd/client/tlsutil"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
 )
 
 const (
 	globalDCLocation     = "global"
 	memberUpdateInterval = time.Minute
 )
 
-// baseClient is a basic client for all other complex client.
-type baseClient struct {
-	urls      atomic.Value // Store as []string
-	clusterID uint64
+type tsoRequest struct {
+	start      time.Time
+	clientCtx  context.Context
+	requestCtx context.Context
+	done       chan error
+	physical   int64
+	logical    int64
+	dcLocation string
+	keyspaceID uint32
+}
+
+// BaseClient defines the general interface for service discovery on a quorum-based cluster
+// or a primary/secondy configured cluster.
+type BaseClient interface {
+	// Init initialize the concrete client underlying
+	Init() error
+	// Close all grpc client connnections
+	CloseClientConns()
+	// GetClusterID returns the ID of the cluster
+	GetClusterID(context.Context) uint64
+	// GetTSOAllocators returns {dc-location -> TSO allocator leader URL} connection map
+	GetTSOAllocators() *sync.Map
+	// GetTSOAllocatorLeaderAddrByDCLocation returns the tso allocator of the given dcLocation
+	GetTSOAllocatorLeaderAddrByDCLocation(dcLocation string) (string, bool)
+	// GetTSOAllocatorClientConnByDCLocation returns the tso allocator grpc client connection
+	// of the given dcLocation
+	GetTSOAllocatorClientConnByDCLocation(dcLocation string) (*grpc.ClientConn, string)
```

Comment:
As a base client interface, could we only provide some general methods to build up the specified client upon it? Introducing TSO-related methods here does not sound good to me.

Reply:
We should move TSO-related methods out, but we can do it incrementally. The reason why we have TSO-related methods here is that previous baseClient(pdBaseClient) mixed up service discovery and tso related service discovery logic. There is historical reason, now we need to refactor it step by step. @JmPotato

Reply:
Ok, if the final goal is like that, then this makes sense to me.

Reply:
BTW, I'll move CreateTsoStream, ProcessTSORequests and the related private functions to another interface. @JmPotato

Reply:
Moved to tso_stream. Template, Abstract Factor, Builder design patterns are used to make it possible.